scrapy/core/scheduler.py

```diff
@@ -1,16 +1,31 @@
-import os
+import abc
+import hashlib
 import json
 import logging
+import os
 from os.path import join, exists
 
-from scrapy.utils.reqser import request_to_dict, request_from_dict
-from scrapy.utils.misc import load_object, create_instance
+from queuelib import RoundRobinQueue
+
+from scrapy.utils.httpobj import urlparse_cached
 from scrapy.utils.job import job_dir
+from scrapy.utils.misc import load_object, create_instance
+from scrapy.utils.reqser import request_to_dict, request_from_dict
 
 logger = logging.getLogger(__name__)
 
 
-class Scheduler(object):
+def _make_file_safe(string):
+    """
+    Make string file safe but readable.
+    """
+    clean_string = "".join([c if c.isalnum() or c in '-._' else '_' for c in string])
+    hash_string = hashlib.md5(string.encode('utf8')).hexdigest()
+    return "{}-{}".format(clean_string[:40], hash_string[:10])
+
+
+class BaseScheduler(object):
+    __metaclass__ = abc.ABCMeta
```

Comment:
this doesn't work in Python 3. Please use [six.with_metaclass](https://pythonhosted.org/six/#six.with_metaclass)